### ISSUE TITLE
`oz compile` will produce a broken artifact at certain `import` statement

### ISSUE THREAD

**Issue by ylv-io** [OpenZeppelin/openzeppelin-sdk#1404]
The following two contracts inside a `contracts` directory will produce a broken `Deep.json` artifact:

Import.sol
```Solidity
pragma solidity ^0.5.0;

import "one/two/Deep.sol";

contract Imports {}
```

Deep.sol
```Solidity
pragma solidity ^0.5.0;

contract Deep {}
```

Key point here is `import "one/two/Deep.sol";` if changed to `import "./one/two/Deep.sol";` will work just fine.

The end of `Deep.json` file:

```JSON
  "bytecode": "0x6080604052348015600f57600080fd5b50603e80601d6000396000f3fe6080604052600080fdfea265627a7a723158207da0e9fc8209e5f15b0ef171c261277950ecb60da823a00a53f338a5ae57b21f64736f6c63430005100032",
  "deployedBytecode": "0x6080604052600080fdfea265627a7a723158207da0e9fc8209e5f15b0ef171c261277950ecb60da823a00a53f338a5ae57b21f64736f6c63430005100032",
  "compiler": {
    "name": "solc",
    "version": "0.5.16+commit.9c3226ce.Emscripten.clang",
    "optimizer": {
      "enabled": false,
      "runs": 200
    },
    "evmVersion": "petersburg"
  }
}
"
  }
}

```
Seems as overwritten by something.

[Full example available.](https://github.com/ylv-io/sdk-compile-bug)

**Comment by spalladino** [comment]
Issue happens because `Deep.sol` is seen as two different contracts with two different paths, and so it's written twice simultaneously:

```js
[ { name: 'Imports',
    fileName: 'Imports.sol',
    sourcePath: 'contracts/Imports.sol' },
  { name: 'Deep',
    fileName: 'Deep.sol',
    sourcePath: 'contracts/one/two/Deep.sol' },
  { name: 'Deep',
    fileName: 'Deep.sol',
    sourcePath: 'one/two/Deep.sol' } ]
```

**Comment by ylv-io** [comment]
Not only its written twice, it's `Deep.json` vary which does cause the error. I think we either should throw an error when devs try to use a path without a leading dot (`one/two/Deep.sol`) to force devs to use `./` or to convert such paths to ones staring with `contracts/...`. WDYT?

**Comment by spalladino** [comment]
The resolver engine does recognize that both imports refer to the same
file, but at some point that info is lost and the project compiler believes
it's two different ones. I'll tackle this one during the week. Thanks for
reporting it!

**Comment by spalladino** [comment]
Potentially related: https://forum.openzeppelin.com/t/unexpected-token-in-json-oz-compile-oz-create-build-bug/1757/6

**Comment by spalladino** [comment]
Been digging further on this (the rabbit hole went deeper than I expected).

When building the input for solc, every file must be identified by a _global name_. Ideally, we want each global name to be the path within the project (`contracts/foo/Bar.sol`) or node modules (`@openzeppelin/contracts/math/SafeMath.sol`). 

When `solc` sees an import statement in a file it's parsing, it runs the following logic:
- If the import path it starts with a dot (ie it's relative to the current file), it resolves the import relative to the current file global name. So, if it finds `./Baz.sol` while parsing `contracts/foo/Bar.sol`, it will yield `contracts/foo/Baz.sol`.
- After resolving it, all folder remappings (if any) are applied. We are not currently using them.
- Finally, it resolves the path is relative to the working directory (or one of the `allow-paths`, this part is still unclear to me).

The problem we have is caused by, as you said @ylv-io, trying to support the `Baz.sol` syntax as an alternative to `./Baz.sol`. Doing so **requires registering `Baz` with `Baz.sol` as global name**, instead of `contracts/foo/Baz.sol`. This potentially causes duplicate entries for `Baz` (as `Baz.sol` and as `contracts/foo/Baz.sol`). Not only that, but now all files imported from `Baz` using `./` syntax need to be registered twice, once for each of Baz's global names.

While we could try some clever usage of remappings to make this work (there could still be ambiguous cases if there are two Solidity files with the same name in different paths), the safest solution would be to just disallow using `Baz.sol` as a replacement of `./Baz.sol`. I'll be modifying the compiler to work that way, which may break a few existing projects, but is much safer.

**Comment by ylv-io** [comment]
Love it. `./Bas.sol` is so much explicit than `Baz.sol`.

**Comment by spalladino** [comment]
Fixed in #1411
